backend/productsearch/api/routes/product.py

```diff
@@ -0,0 +1,148 @@
+import asyncio
+
+import numpy as np
+from fastapi import APIRouter, Depends
+from redis.commands.search.document import Document
+from redis.commands.search.query import Query
+from redisvl.index import AsyncSearchIndex
+from redisvl.query import FilterQuery, VectorQuery
+from redisvl.query.filter import FilterExpression, Tag
+
+from productsearch import config
+from productsearch.api.schema.product import (
+    ProductSearchResponse,
+    ProductVectorSearchResponse,
+    SimilarityRequest,
+)
+from productsearch.db import redis_helpers
+
+
+router = APIRouter()
+
+
+def create_count_query(filter_expression: FilterExpression) -> Query:
+    """
+    Create a "count" query where simply want to know how many records
+    match a particular filter expression
+
+    Args:
+        filter_expression (FilterExpression): The filter expression for the query.
+
+    Returns:
+        Query: The Redis query object.
+    """
+    return Query(str(filter_expression)).no_content().dialect(2)
+
+
+@router.get(
+    "/",
+    response_model=ProductSearchResponse,
+    name="product:get_product_samples",
+    operation_id="get_products_samples",
+)
+async def get_products(
+    limit: int = 20,
+    skip: int = 0,
+    gender: str = "",
+    category: str = "",
+    index: AsyncSearchIndex = Depends(redis_helpers.get_async_index),
+):
+    """_summary_
+
+    Args:
+        limit (int, optional): _description_. Defaults to 20.
+        skip (int, optional): _description_. Defaults to 0.
+        gender (str, optional): _description_. Defaults to "".
+        category (str, optional): _description_. Defaults to "".
+
+    Returns:
+        ProductSearchResponse model
+    """
+    # Build query
+    filter_expression = (Tag("gender") == gender) & (Tag("category") == category)
+    filter_query = FilterQuery(return_fields=[], filter_expression=filter_expression)
+    # Execute search
+    result_papers = await index.search(filter_query.query.paging(skip, limit))
+    # return prepare_response(result_papers.total, result_papers)
+    return ProductSearchResponse(
+        total=result_papers.total, products=[d.__dict__ for d in result_papers.docs]
+    )
+
+
+@router.post(
+    "/vectorsearch/image",
+    response_model=ProductVectorSearchResponse,
+    name="product:find_similar_by_image",
+    operation_id="compute_image_similarity",
+)
+async def find_products_by_image(
+    similarity_request: SimilarityRequest,
+    index: AsyncSearchIndex = Depends(redis_helpers.get_async_index),
+) -> ProductVectorSearchResponse:
+
+    # Fetch paper key and the vector from the HASH, cast to numpy array
+    product_key = f"{index.schema.index.prefix}:{similarity_request.product_id}"
```

Comment:
We can do this slightly more elegantly using redisvl.

`index.fetch(similarity_request.product_id)` --> will pull back the Hash. Then you can grab the `img_vector` field and cast back to an array